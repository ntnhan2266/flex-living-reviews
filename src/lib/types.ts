export interface HostawayReview {
  id: number;
  type: string;
  status: string;
  rating: number | null;
  publicReview: string;
  reviewCategory: {
    category: string;
    rating: number;
  }[];
  submittedAt: string;
  guestName: string;
  listingName: string;
}

export type Channel = 'hostaway' | 'airbnb' | 'booking' | 'google';
export type ReviewStatus =
  | 'awaiting'
  | 'pending'
  | 'scheduled'
  | 'submitted'
  | 'published'
  | 'expired'
  | 'approved'
  | 'public';

export interface NormalizedReview {
  id: string;
  propertyName: string;
  guestName: string;
  rating: number;
  comment: string;
  date: Date;
  channel: Channel;
  categories: Record<string, number | undefined>;
  status: ReviewStatus;
}

export interface PropertyStats {
  id: string;
  name: string;
  averageRating: number;
  totalReviews: number;
  approvedReviews: number;
  recentTrend: 'up' | 'down' | 'stable';
}

export type CategoryKey =
  | 'cleanliness'
  | 'communication'
  | 'location'
  | 'accuracy'
  | 'checkin'
  | 'value';

export type SortKey = 'date_desc' | 'date_asc' | 'rating_desc' | 'rating_asc';

export type Filters = {
  property: string;
  channel: '' | 'hostaway' | 'airbnb' | 'booking' | 'google';
  rating: '' | '1' | '2' | '3' | '4' | '5';
  startDate: string; // yyyy-mm-dd
  endDate: string;   // yyyy-mm-dd
  status: 'all' | 'pending' | 'approved';
  category?: '' | CategoryKey;
  minCategory?: '' | '1' | '2' | '3' | '4' | '5';
  sort?: SortKey;
};

export interface ApiResponse<T> {
  status: 'success' | 'error';
  data?: T;
  message?: string;
  total?: number;
}
