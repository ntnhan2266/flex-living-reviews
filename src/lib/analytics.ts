// src/lib/analytics.ts
import { NormalizedReview, ReviewStatus } from '@/lib/types';

export type Granularity = 'day' | 'week' | 'month';

export type TimeseriesDatum = {
  date: string; // "YYYY-MM-DD" for day/week start/month start
  count: number; // number of reviews in bucket
  avgRating: number; // 0..5 average in bucket (0 when no ratings)
};

/**
 * Buckets a JS Date into an ISO key by granularity:
 * - day  -> YYYY-MM-DD
 * - week -> YYYY-MM-DD of the week's Monday
 * - month-> YYYY-MM-01
 */
export function bucketKey(d: Date, granularity: Granularity): string {
  const dt = new Date(d);
  dt.setHours(0, 0, 0, 0);

  if (granularity === 'week') {
    // Monday as week start
    const day = dt.getDay() || 7; // 1..7 (Mon..Sun)
    if (day !== 1) dt.setDate(dt.getDate() - (day - 1));
  } else if (granularity === 'month') {
    dt.setDate(1);
  }

  // fastest way to get YYYY-MM-DD without deps
  return dt.toISOString().slice(0, 10);
}

/**
 * Build a timeseries for reviews.
 * @param reviews - list of reviews
 * @param options.statuses - optional statuses to include (e.g. ['approved'])
 * @param options.granularity - day|week|month (default 'day')
 */
export function getReviewTimeseries(
  reviews: NormalizedReview[],
  options?: { statuses?: ReviewStatus[]; granularity?: Granularity },
): TimeseriesDatum[] {
  const allowed = new Set(options?.statuses ?? []);
  const gran = options?.granularity ?? 'day';

  const grouped: Record<string, { count: number; sum: number; withRating: number }> = {};

  for (const r of reviews) {
    if (allowed.size && !allowed.has(r.status)) continue;

    // normalize date (handle string | Date)
    const d = r.date instanceof Date ? r.date : new Date(r.date);
    const key = bucketKey(d, gran);
    if (!grouped[key]) grouped[key] = { count: 0, sum: 0, withRating: 0 };

    grouped[key].count += 1;

    const val = Number(r.rating);
    if (Number.isFinite(val)) {
      grouped[key].sum += val;
      grouped[key].withRating += 1;
    }
  }

  return Object.entries(grouped)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
    .map(([date, { count, sum, withRating }]) => ({
      date,
      count,
      avgRating: withRating ? Number((sum / withRating).toFixed(2)) : 0,
    }));
}
