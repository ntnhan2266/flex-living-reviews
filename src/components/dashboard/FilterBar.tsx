import { Filter, X } from 'lucide-react';
import { Filters } from '@/lib/types';

const CATEGORY_OPTIONS: Array<{ label: string; value: NonNullable<Filters['category']> }> = [
  { label: 'Cleanliness', value: 'cleanliness' },
  { label: 'Communication', value: 'communication' },
  { label: 'Location', value: 'location' },
  { label: 'Accuracy', value: 'accuracy' },
  { label: 'Check-in', value: 'checkin' },
  { label: 'Value', value: 'value' },
];

export default function FilterBar({
  filters,
  onFiltersChange,
}: {
  filters: Filters;
  onFiltersChange: (f: Filters) => void;
}) {
  const clearFilters = () =>
    onFiltersChange({
      property: '',
      channel: '',
      rating: '',
      startDate: '',
      endDate: '',
      status: 'all',
      category: '',
      minCategory: '',
      sort: 'date_desc',
    });

  const hasActive = Object.entries(filters).some(
    ([k, v]) =>
      // treat defaults as inactive
      (v && v !== 'all') &&
      !(
        (k === 'sort' && v === 'date_desc') ||
        (k === 'category' && v === '') ||
        (k === 'minCategory' && v === '')
      )
  );

  return (
    <div className="bg-white p-4 rounded-lg shadow-sm border border-gray-200">
      <div className="flex items-center gap-4 flex-wrap">
        <div className="flex items-center gap-2">
          <Filter className="w-4 h-4 text-gray-900" aria-hidden />
          <span className="text-sm font-medium text-gray-900">Filters</span>
        </div>

        {/* Channel */}
        <label className="sr-only" htmlFor="channel">Channel</label>
        <select
          id="channel"
          value={filters.channel}
          onChange={(e) => onFiltersChange({ ...filters, channel: e.target.value as Filters['channel'] })}
          className="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        >
          <option value="">All channels</option>
          <option value="hostaway">Hostaway</option>
          <option value="airbnb">Airbnb</option>
          <option value="booking">Booking.com</option>
          <option value="google">Google</option>
        </select>

        {/* Min overall rating */}
        <label className="sr-only" htmlFor="rating">Min rating</label>
        <select
          id="rating"
          value={filters.rating}
          onChange={(e) => onFiltersChange({ ...filters, rating: e.target.value as Filters['rating'] })}
          className="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        >
          <option value="">All ratings</option>
          <option value="5">5 Stars</option>
          <option value="4">4+ Stars</option>
          <option value="3">3+ Stars</option>
          <option value="2">2+ Stars</option>
          <option value="1">1+ Stars</option>
        </select>

        {/* Status */}
        <label className="sr-only" htmlFor="status">Status</label>
        <select
          id="status"
          value={filters.status}
          onChange={(e) => onFiltersChange({ ...filters, status: e.target.value as Filters['status'] })}
          className="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        >
          <option value="all">All reviews</option>
          <option value="pending">Pending</option>
          <option value="approved">Approved</option>
          <option value="rejected">Rejected</option>
        </select>

        {/* NEW: Category */}
        <label className="sr-only" htmlFor="category">Category</label>
        <select
          id="category"
          value={filters.category ?? ''}
          onChange={(e) => onFiltersChange({ ...filters, category: e.target.value as NonNullable<Filters['category']> })}
          className="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        >
          <option value="">All categories</option>
          {CATEGORY_OPTIONS.map((opt) => (
            <option key={opt.value} value={opt.value}>{opt.label}</option>
          ))}
        </select>

        {/* NEW: Min sub-score for that category */}
        <label className="sr-only" htmlFor="minCategory">Min sub-score</label>
        <select
          id="minCategory"
          value={filters.minCategory ?? ''}
          onChange={(e) => onFiltersChange({ ...filters, minCategory: e.target.value as NonNullable<Filters['minCategory']> })}
          className="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
          disabled={!filters.category}
          title={filters.category ? 'Minimum sub-score for selected category' : 'Select a category first'}
        >
          <option value="">Any</option>
          <option value="5">5</option>
          <option value="4">4+</option>
          <option value="3">3+</option>
          <option value="2">2+</option>
          <option value="1">1+</option>
        </select>

        {/* NEW: Sort */}
        <label className="sr-only" htmlFor="sort">Sort</label>
        <select
          id="sort"
          value={filters.sort ?? 'date_desc'}
          onChange={(e) => onFiltersChange({ ...filters, sort: e.target.value as NonNullable<Filters['sort']> })}
          className="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        >
          <option value="date_desc">Newest</option>
          <option value="date_asc">Oldest</option>
          <option value="rating_desc">Rating: High → Low</option>
          <option value="rating_asc">Rating: Low → High</option>
        </select>

        {/* Date range */}
        <div className="flex items-center gap-2">
          <label className="sr-only" htmlFor="startDate">Start date</label>
          <input
            id="startDate"
            type="date"
            value={filters.startDate}
            onChange={(e) => onFiltersChange({ ...filters, startDate: e.target.value })}
            className="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
          />
          <span className="text-gray-400">to</span>
          <label className="sr-only" htmlFor="endDate">End date</label>
          <input
            id="endDate"
            type="date"
            value={filters.endDate}
            onChange={(e) => onFiltersChange({ ...filters, endDate: e.target.value })}
            className="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
          />
        </div>

        <button
          onClick={clearFilters}
          disabled={!hasActive}
          className="flex items-center gap-1 px-3 py-1 text-sm text-red-600 hover:bg-red-50 disabled:opacity-50 disabled:cursor-not-allowed rounded-md transition-colors"
        >
          <X className="w-4 h-4" aria-hidden />
          Clear
        </button>
      </div>
    </div>
  );
}
