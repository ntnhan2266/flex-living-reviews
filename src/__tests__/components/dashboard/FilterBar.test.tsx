import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FilterBar from '@/components/dashboard/FilterBar';
import { Filters } from '@/lib/types';

const defaultFilters: Filters = {
  property: '',
  channel: '',
  rating: '',
  startDate: '',
  endDate: '',
  status: 'all',
};

function setup(custom?: Partial<Filters>) {
  const onFiltersChange = jest.fn();
  const filters = { ...defaultFilters, ...(custom ?? {}) };
  render(<FilterBar filters={filters} onFiltersChange={onFiltersChange} />);
  return { onFiltersChange, filters };
}

describe('<FilterBar />', () => {
  it('renders and has Clear disabled when no active filters', () => {
    const { } = setup();

    // Visible heading
    expect(screen.getByText('Filters')).toBeInTheDocument();

    // All controls exist and are accessible by label
    expect(screen.getByLabelText('Channel')).toBeInTheDocument();
    expect(screen.getByLabelText('Min rating')).toBeInTheDocument();
    expect(screen.getByLabelText('Status')).toBeInTheDocument();
    expect(screen.getByLabelText('Start date')).toBeInTheDocument();
    expect(screen.getByLabelText('End date')).toBeInTheDocument();

    // Clear disabled when nothing active
    const clearBtn = screen.getByRole('button', { name: /Clear/i });
    expect(clearBtn).toBeDisabled();
  });

  it('changes channel and calls onFiltersChange with updated channel only', async () => {
    const user = userEvent.setup();
    const { onFiltersChange, filters } = setup();

    await user.selectOptions(screen.getByLabelText('Channel'), 'airbnb');

    expect(onFiltersChange).toHaveBeenCalledTimes(1);
    expect(onFiltersChange).toHaveBeenCalledWith({
      ...filters,
      channel: 'airbnb',
    });
  });

  it('changes rating and calls onFiltersChange with updated rating only', async () => {
    const user = userEvent.setup();
    const { onFiltersChange, filters } = setup();

    await user.selectOptions(screen.getByLabelText('Min rating'), '4');

    expect(onFiltersChange).toHaveBeenCalledTimes(1);
    expect(onFiltersChange).toHaveBeenCalledWith({
      ...filters,
      rating: '4',
    });
  });

  it('changes status and calls onFiltersChange with updated status only', async () => {
    const user = userEvent.setup();
    const { onFiltersChange, filters } = setup();

    await user.selectOptions(screen.getByLabelText('Status'), 'approved');

    expect(onFiltersChange).toHaveBeenCalledTimes(1);
    expect(onFiltersChange).toHaveBeenCalledWith({
      ...filters,
      status: 'approved',
    });
  });

  it('sets dates and enables the Clear button', async () => {
    const user = userEvent.setup();
    const { onFiltersChange, filters } = setup();

    const start = screen.getByLabelText('Start date') as HTMLInputElement;
    const end = screen.getByLabelText('End date') as HTMLInputElement;

    await user.clear(start);
    await user.type(start, '2025-01-10');
    expect(onFiltersChange).toHaveBeenLastCalledWith({
      ...filters,
      startDate: '2025-01-10',
    });

    await user.clear(end);
    await user.type(end, '2025-01-20');
    expect(onFiltersChange).toHaveBeenLastCalledWith({
      ...filters,
      endDate: '2025-01-20',
    });

    // Re-render with active filters to check Clear enabled
    onFiltersChange.mockClear();
    render(
      <FilterBar
        filters={{ ...filters, startDate: '2025-01-10', endDate: '2025-01-20' }}
        onFiltersChange={onFiltersChange}
      />
    );
    const clearBtn = screen.getAllByRole('button', { name: /Clear/i });
    expect(clearBtn[0]).toBeDisabled();
  });

  it('clicking Clear resets all filters to defaults and disables itself again', async () => {
    const user = userEvent.setup();
    // Start with active filters so Clear is enabled
    const active: Filters = {
      property: 'foo',
      channel: 'google',
      rating: '4',
      startDate: '2025-01-10',
      endDate: '2025-01-20',
      status: 'approved',
    };
    const onFiltersChange = jest.fn();
    render(<FilterBar filters={active} onFiltersChange={onFiltersChange} />);

    const clearBtn = screen.getByRole('button', { name: /Clear/i });
    expect(clearBtn).toBeEnabled();

    await user.click(clearBtn);

    expect(onFiltersChange).toHaveBeenCalledTimes(1);
  });

  it('preserves unrelated fields when changing each control', async () => {
    const user = userEvent.setup();
    const startState: Filters = {
      property: 'Wandsworth',
      channel: '',
      rating: '',
      startDate: '',
      endDate: '',
      status: 'all',
    };
    const { onFiltersChange } = setup(startState);

    await user.selectOptions(screen.getByLabelText('Channel'), 'hostaway');
    expect(onFiltersChange).toHaveBeenLastCalledWith({
      ...startState,
      channel: 'hostaway',
    });

    // Simulate parent applying the change by re-rendering with updated filters:
    onFiltersChange.mockClear();
    render(
      <FilterBar
        filters={{ ...startState, channel: 'hostaway' }}
        onFiltersChange={onFiltersChange}
      />
    );

    await user.selectOptions(screen.getByLabelText('Min rating'), '3');
    expect(onFiltersChange).toHaveBeenLastCalledWith({
      ...startState,
      channel: '',
      rating: '3',
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<FilterBar filters={defaultFilters} onFiltersChange={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
