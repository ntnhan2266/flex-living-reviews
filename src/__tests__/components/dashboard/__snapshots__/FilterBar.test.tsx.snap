// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<FilterBar /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white p-4 rounded-lg shadow-sm border border-gray-200"
  >
    <div
      class="flex items-center gap-4 flex-wrap"
    >
      <div
        class="flex items-center gap-2"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-funnel w-4 h-4 text-gray-900"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 20a1 1 0 0 0 .553.895l2 1A1 1 0 0 0 14 21v-7a2 2 0 0 1 .517-1.341L21.74 4.67A1 1 0 0 0 21 3H3a1 1 0 0 0-.742 1.67l7.225 7.989A2 2 0 0 1 10 14z"
          />
        </svg>
        <span
          class="text-sm font-medium text-gray-900"
        >
          Filters
        </span>
      </div>
      <label
        class="sr-only"
        for="channel"
      >
        Channel
      </label>
      <select
        class="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        id="channel"
      >
        <option
          value=""
        >
          All channels
        </option>
        <option
          value="hostaway"
        >
          Hostaway
        </option>
        <option
          value="airbnb"
        >
          Airbnb
        </option>
        <option
          value="booking"
        >
          Booking.com
        </option>
        <option
          value="google"
        >
          Google
        </option>
      </select>
      <label
        class="sr-only"
        for="rating"
      >
        Min rating
      </label>
      <select
        class="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        id="rating"
      >
        <option
          value=""
        >
          All ratings
        </option>
        <option
          value="5"
        >
          5 Stars
        </option>
        <option
          value="4"
        >
          4+ Stars
        </option>
        <option
          value="3"
        >
          3+ Stars
        </option>
        <option
          value="2"
        >
          2+ Stars
        </option>
        <option
          value="1"
        >
          1+ Stars
        </option>
      </select>
      <label
        class="sr-only"
        for="status"
      >
        Status
      </label>
      <select
        class="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        id="status"
      >
        <option
          value="all"
        >
          All reviews
        </option>
        <option
          value="pending"
        >
          Pending
        </option>
        <option
          value="approved"
        >
          Approved
        </option>
        <option
          value="rejected"
        >
          Rejected
        </option>
      </select>
      <label
        class="sr-only"
        for="category"
      >
        Category
      </label>
      <select
        class="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        id="category"
      >
        <option
          value=""
        >
          All categories
        </option>
        <option
          value="cleanliness"
        >
          Cleanliness
        </option>
        <option
          value="communication"
        >
          Communication
        </option>
        <option
          value="location"
        >
          Location
        </option>
        <option
          value="accuracy"
        >
          Accuracy
        </option>
        <option
          value="checkin"
        >
          Check-in
        </option>
        <option
          value="value"
        >
          Value
        </option>
      </select>
      <label
        class="sr-only"
        for="minCategory"
      >
        Min sub-score
      </label>
      <select
        class="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        disabled=""
        id="minCategory"
        title="Select a category first"
      >
        <option
          value=""
        >
          Any
        </option>
        <option
          value="5"
        >
          5
        </option>
        <option
          value="4"
        >
          4+
        </option>
        <option
          value="3"
        >
          3+
        </option>
        <option
          value="2"
        >
          2+
        </option>
        <option
          value="1"
        >
          1+
        </option>
      </select>
      <label
        class="sr-only"
        for="sort"
      >
        Sort
      </label>
      <select
        class="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
        id="sort"
      >
        <option
          value="date_desc"
        >
          Newest
        </option>
        <option
          value="date_asc"
        >
          Oldest
        </option>
        <option
          value="rating_desc"
        >
          Rating: High → Low
        </option>
        <option
          value="rating_asc"
        >
          Rating: Low → High
        </option>
      </select>
      <div
        class="flex items-center gap-2"
      >
        <label
          class="sr-only"
          for="startDate"
        >
          Start date
        </label>
        <input
          class="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
          id="startDate"
          type="date"
          value=""
        />
        <span
          class="text-gray-400"
        >
          to
        </span>
        <label
          class="sr-only"
          for="endDate"
        >
          End date
        </label>
        <input
          class="border border-gray-300 text-gray-700 rounded-md px-3 py-1 text-sm focus:outline-none focus:ring-2 focus:ring-blue-500"
          id="endDate"
          type="date"
          value=""
        />
      </div>
      <button
        class="flex items-center gap-1 px-3 py-1 text-sm text-red-600 hover:bg-red-50 disabled:opacity-50 disabled:cursor-not-allowed rounded-md transition-colors"
        disabled=""
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-x w-4 h-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
        Clear
      </button>
    </div>
  </div>
</DocumentFragment>
`;
